File: app/services/auth_service.py
# app/services/auth_service.py
import jwt
import time
import os
from datetime import datetime, timedelta

JWT_SECRET = os.getenv('JWT_SECRET', 'fallback_super_secret_key')
INVALIDATED_TOKENS = set()
LOGIN_ATTEMPTS = {}
MAX_ATTEMPTS = 5
LOCKOUT_TIME = 300

def _generate_token(user_id, token_type='access'):
    expiry = datetime.utcnow() + timedelta(minutes=15 if token_type == 'access' else days=7)
    payload = {
        'user_id': user_id,
        'exp': expiry,
        'iat': datetime.utcnow(),
        'type': token_type
    }
    return jwt.encode(payload, JWT_SECRET, algorithm='HS256')

def login_user(username, password):
    user_id = 1

    if username in LOGIN_ATTEMPTS and LOGIN_ATTEMPTS[username]['attempts'] >= MAX_ATTEMPTS:
        if (time.time() - LOGIN_ATTEMPTS[username]['last_attempt']) < LOCKOUT_TIME:
            print(f"Account {username} locked out.")
            return None, None
        else:
            LOGIN_ATTEMPTS[username] = {'attempts': 0, 'last_attempt': time.time()}

    if password == 'secure_password':
        access_token = _generate_token(user_id, 'access')
        refresh_token = _generate_token(user_id, 'refresh')
        LOGIN_ATTEMPTS.pop(username, None)
        return access_token, refresh_token
    else:
        LOGIN_ATTEMPTS.setdefault(username, {'attempts': 0, 'last_attempt': time.time()})['attempts'] += 1
        LOGIN_ATTEMPTS[username]['last_attempt'] = time.time()
        print(f"Login failed for {username}. Attempts: {LOGIN_ATTEMPTS[username]['attempts']}")
        return None, None

def refresh_access_token(old_refresh_token):
    try:
        payload = jwt.decode(old_refresh_token, JWT_SECRET, algorithms=['HS256'])
        if payload['type'] != 'refresh' or old_refresh_token in INVALIDATED_TOKENS:
            raise jwt.InvalidTokenError

        INVALIDATED_TOKENS.add(old_refresh_token)
        new_access_token = _generate_token(payload['user_id'], 'access')
        new_refresh_token = _generate_token(payload['user_id'], 'refresh')
        return new_access_token, new_refresh_token
    except jwt.ExpiredSignatureError:
        print("Refresh token expired")
        return None, None
    except jwt.InvalidTokenError:
        print("Invalid refresh token")
        return None, None

def invalidate_session(access_token_or_refresh_token):
    INVALIDATED_TOKENS.add(access_token_or_refresh_token)
    print("Token invalidated.")

File: src/services/authService.js
console.log('// src/services/authService.js');
const jwt = require('jsonwebtoken');
const JWT_SECRET = process.env.JWT_SECRET || 'fallback_super_secret_key';
const INVALIDATED_TOKENS = new Set();
const LOGIN_ATTEMPTS = {};
const MAX_ATTEMPTS = 5;
const LOCKOUT_TIME = 300 * 1000;

function _generateToken(userId, tokenType = 'access') {
  const expiresIn = tokenType === 'access' ? '15m' : '7d';
  return jwt.sign({ userId, type: tokenType }, JWT_SECRET, { expiresIn });
}

function loginUser(username, password) {
  const userId = 1;

  if (LOGIN_ATTEMPTS[username] && LOGIN_ATTEMPTS[username].attempts >= MAX_ATTEMPTS) {
    if ((Date.now() - LOGIN_ATTEMPTS[username].lastAttempt) < LOCKOUT_TIME) {
      console.log(`Account ${username} locked out.`);
      return { accessToken: null, refreshToken: null };
    } else {
      LOGIN_ATTEMPTS[username] = { attempts: 0, lastAttempt: Date.now() };
    }
  }

  if (password === 'secure_password') {
    const accessToken = _generateToken(userId, 'access');
    const refreshToken = _generateToken(userId, 'refresh');
    delete LOGIN_ATTEMPTS[username];
    return { accessToken, refreshToken };
  } else {
    LOGIN_ATTEMPTS[username] = LOGIN_ATTEMPTS[username] || { attempts: 0, lastAttempt: Date.now() };
    LOGIN_ATTEMPTS[username].attempts++;
    LOGIN_ATTEMPTS[username].lastAttempt = Date.now();
    console.log(`Login failed for ${username}. Attempts: ${LOGIN_ATTEMPTS[username].attempts}`);
    return { accessToken: null, refreshToken: null };
  }
}

function refreshAccessToken(oldRefreshToken) {
  try {
    const payload = jwt.verify(oldRefreshToken, JWT_SECRET);
    if (payload.type !== 'refresh' || INVALIDATED_TOKENS.has(oldRefreshToken)) {
      throw new Error('Invalid refresh token');
    }

    INVALIDATED_TOKENS.add(oldRefreshToken);
    const newAccessToken = _generateToken(payload.userId, 'access');
    const newRefreshToken = _generateToken(payload.userId, 'refresh');
    return { accessToken: newAccessToken, refreshToken: newRefreshToken };
  } catch (error) {
    console.error(`Refresh token error: ${error.message}`);
    return { accessToken: null, refreshToken: null };
  }
}

function invalidateSession(token) {
  INVALIDATED_TOKENS.add(token);
  console.log('Token invalidated.');
}

module.exports = { loginUser, refreshAccessToken, invalidateSession };
